### ISSUE TITLE
Add stride operation primitive

### ISSUE THREAD

**Issue by riccardofelluga** [Lightning-AI/lightning-thunder#135]
## 🚀 Feature

I would like to have Thunder manage stride information to allow for tensor manipulations. 

In particular I think the following points need to be discussed:

- Where does the stride information go? Should it be part of `TensorProxy`?
- What can a stride manipulation primitive look like? Are there any particular things we need to be careful about?

### Motivation

This will enable us to add new operators that reshape the tensor using the stride like `torch.as_strided` or `torch.Tensor.unfold`.

**Comment by mruberry** [comment]
fyi @jjsjann123 

We have a stride order primitive today:

https://github.com/Lightning-AI/lightning-thunder/blob/393828fa1dd45b3a9ff695f1bdbf005f5a669e5e/thunder/core/prims.py#L158

We don't have plans to implement support for setting exact strides, just to reorder them (like the stride order primitive does)

**Comment by riccardofelluga** [comment]
Hey! So what I was thinking was more about setting exact strides.

However, @IvanYashchuk pointed out to me that this discussion has been held once before. That discussion led to the following conclusion: 
> [...] we did identify that having a stride semantics in the graph is useful, but it's very hard to do it right and efficiently, without putting unnecessary constraints on backends.

Additionally, there is a [comment on jax](https://github.com/google/jax/issues/3171#issuecomment-631826501) that I think summarizes enough the possible complexity of having such a feature.

In conclusion, I think it's fair to say that unless there is a stronger need for it, it's better to put this idea aside for now
